kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,426 @@
+            Some(filter) => filter.apply(actions.as_ref())?,
+            None => vec![true; actions.len()],
+        };
+
+        let mut visitor = PreparePhaseVisitor::new(
+            selection_vector,
+            &mut add_paths,
+            &mut remove_dvs,
+            &mut has_cdc_action,
+            &mut timestamp,
+        );
+        visitor.visit_rows_of(actions.as_ref())?;
+
+        if let Some(protocol) = visitor.protocol {
+            protocol.ensure_read_supported()?;
+        }
+        if let Some((schema, configuration)) = visitor.metadata_info {
+            let schema: StructType = serde_json::from_str(&schema)?;
+            // Currently, schema compatibility is defined as having equal schema types. In the
+            // future, more permisive schema evolution will be supported.
+            // See: https://github.com/delta-io/delta-kernel-rs/issues/523
+            require!(
+                table_schema.as_ref() == &schema,
+                Error::change_data_feed_incompatible_schema(table_schema, &schema)
+            );
+            let table_properties = TableProperties::from(configuration);
+            require!(
+                table_properties.enable_change_data_feed.unwrap_or(false),
+                Error::change_data_feed_unsupported(commit_file.version)
+            )
+        }
+    }
+    // We resolve the remove deletion vector map after visiting the entire commit.
+    if has_cdc_action {
+        remove_dvs.clear();
+    } else {
+        // The only (path, deletion_vector) pairs we must track are ones whose path is the
+        // same as an `add` action.
+        remove_dvs.retain(|rm_path, _| add_paths.contains(rm_path));
+    }
+    Ok(LogReplayScanner {
+        timestamp,
+        commit_file,
+        has_cdc_action,
+        remove_dvs,
+        filter,
+    })
+}
+impl LogReplayScanner {
+    fn into_scan_batches(
+        self,
+        engine: Arc<dyn Engine>,
+    ) -> DeltaResult<impl Iterator<Item = DeltaResult<TableChangesScanData>>> {
+        let Self {
+            has_cdc_action,
+            remove_dvs,
+            commit_file,
+            // TODO: Add the timestamp as a column with an expression
+            timestamp: _,
+            filter,
+        } = self;
+        let remove_dvs = Arc::new(remove_dvs);
+
+        let schema = FileActionSelectionVisitor::schema()?;
+        let action_iter = engine.get_json_handler().read_json_files(
+            &[commit_file.location.clone()],
+            schema,
+            None,
+        )?;
+
+        let result = action_iter.map(move |actions| -> DeltaResult<_> {
+            let actions = actions?;
+
+            // Apply data skipping to get back a selection vector for actions that passed skipping.
+            // We start our selection vector based on what was filtered. We will add to this vector
+            // below if a file has been removed. Note: None implies all files passed data skipping.
```

Comment:
Rescuing https://github.com/delta-io/delta-kernel-rs/pull/540#discussion_r1866676550?
> I don't see us adding to the filter anywhere? (it's not declared mut for example)?
>
> Also -- should we save the filter and reuse it in the scan phase, instead of re-running the predicate?

Reply:
This might be a copy/pasta issue from the `scan/log_replay.rs` code

Reply:
ah true. This is copy/pasted. To my understanding it still holds since we modify the selection vector in the `FileActionSelectionVisitor`.

Reply:
... but then we throw it away and phase 2 has to start over? That seems non-ideal?

Reply:
Based on the effectivization discussion, we want to avoid data skipping filter in 1st phase right?